### PR TITLE
WIP: Testing from Saturday

### PR DESCRIPTION
**Pull request by roshanr10**
All unmerged changes from today– need to finish up, review and merge to testing

### CODE REVIEW

File: src/main/java/frc/team4909/robot/RobotConstants.java [hunk 1]
```diff
@@ -20,11 +20,11 @@
   public static final double cargoIntakeHoldSpeed = 0;
   public static final double cargoIntakeOutSpeed = -1.0;
 
-  public static final double cargoIntakeCurrentLimit = 20;
+  public static final double cargoIntakeCurrentLimit = 0;
```
Comment (derrell) [inline src/main/java/frc/team4909/robot/RobotConstants.java]: Doesn't a limit of 0 mean that it should be cutting off if any current is drawn? This seems off...

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java [hunk 2]
```diff
@@ -15,26 +15,32 @@
...
 
         //Lift
-        masterSRX = new WPI_TalonSRX(RobotConstants.elevatorSRXID);  //master SRX
-        VictorSPX1 = new WPI_VictorSPX(RobotConstants.elevatorSPX1ID); //slave SPX 1
-        VictorSPX2 = new WPI_VictorSPX(RobotConstants.elevatorSPX2ID); //slave SPX 2
-        VictorSPX3 = new WPI_VictorSPX(RobotConstants.elevatorSPX3ID); //slave SPX 3
+        leftSRX = new WPI_TalonSRX(RobotMap.elevatorSRXID);  //master SRX
+        leftSPX = new WPI_VictorSPX(RobotMap.elevatorSPX1ID); //slave SPX 1
+        rightSRX1 = new WPI_VictorSPX(RobotMap.elevatorSPX2ID); //slave SPX 2
+        rightSRX2 = new WPI_VictorSPX(RobotMap.elevatorSPX3ID); //slave SPX 3
```
Comment (derrell) [inline src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: These two "right" variables appear to be named incorrectly. The controller is SPX but the variable says it's SRX. I think these variable names should be `rightSPX1` and `rightSPX2`

File: src/main/java/frc/team4909/robot/subsystems/intake/CargoIntakeIn.java [hunk 3]
```diff
@@ -15,6 +15,6 @@ protected void execute() {
 
     @Override
     protected boolean isFinished() {
-        return Robot.intakeSubsystem.getCargoIntakeCurrent() > RobotConstants.cargoIntakeCurrentLimit;
+        return false;
```
Comment (derrell) [inline src/main/java/frc/team4909/robot/subsystems/intake/CargoIntakeIn.java]: Never stop? What stops it? And if it is intended to not be stopped here, should it be an InstanceCommand rather than a Command?

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java (hunk 2 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: yep, good catch, will fix

File: src/main/java/frc/team4909/robot/RobotConstants.java (hunk 1 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/RobotConstants.java]: Yeah, this was last minute testing, I don't think the limiting is working as intended. This should've cut off, but it did not... this was the last code pushed to the robot right before we were kicked out of the shop.

File: src/main/java/frc/team4909/robot/subsystems/intake/CargoIntakeIn.java (hunk 3 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/subsystems/intake/CargoIntakeIn.java]: we're still trying to correct this logic, as it's not working as intended– this was also part of last minute testing

**Comment by roshanr10** [comment]
Just replied to comments, this is nowhere near ready to merge, this is just the current state of robot code after our testing and communication with drive team.

We just wanted to open the PR so it was more easily accessible for others to see since @ashwinc12 and I will be a bit late for tomorrows meeting.

**Comment by derrell** [comment]
Yup, I saw that it was WIP. Didn't know quite what state it was, though, so commented on it as if it was close to ready.

File: src/main/java/frc/team4909/robot/Robot.java [hunk 4]
```diff
@@ -14,10 +16,17 @@
 import frc.team4909.robot.subsystems.drivetrain.DriveTrainSubsystem;
 import frc.team4909.robot.subsystems.drivetrain.commands.InvertDriveDirection;
 import frc.team4909.robot.subsystems.intake.IntakeSubsystem;
-import frc.team4909.robot.subsystems.elevator.ElevatorSubsystem;;
+import frc.team4909.robot.subsystems.elevator.ElevatorSubsystem;
+import frc.team4909.robot.subsystems.elevator.commands.SetElevatorPosition;
+import frc.team4909.robot.subsystems.elevatorarm.ElevatorArmSubsystem;
+import frc.team4909.robot.subsystems.elevatorarm.ElevatorArmSubsystem;
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/Robot.java]: why is this here twice?

File: src/main/java/frc/team4909/robot/Robot.java [hunk 5]
```diff
@@ -40,43 +50,50 @@
...
+    
 
-    driverGamepad = new BionicF310(RobotConstants.driverGamepadPort, // Port
+
+    driverGamepad = new BionicF310(RobotMap.driverGamepadPort, // Port
                                    RobotConstants.driverGamepadDeadzone, // Deadzone
                                    RobotConstants.driverGamepadSensitivity // Gamepad sensitivity
     );
 
-    manipulatorGamepad = new  BionicF310(RobotConstants.manipulatorGamepadPort, // Port
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/Robot.java]: just curious, why was this switched to constants and maps?

Comment (roshanr10) [reply src/main/java/frc/team4909/robot/Robot.java]: it's fine either way, I'm just going thru the whole branch rn

File: src/main/java/frc/team4909/robot/Robot.java [hunk 6]
```diff
@@ -40,43 +50,50 @@
...
+    
     driverGamepad.buttonPressed(BionicF310.A, new Linefollow());
-    driverGamepad.buttonPressed(BionicF310.X, new CargoIntakeIn());
-    driverGamepad.buttonPressed(BionicF310.Y, new CargoIntakeOut());
-    driverGamepad.buttonPressed(BionicF310.LB, new HatchPanelIntakeOpen());
-    driverGamepad.buttonPressed(BionicF310.RB, new HatchPanelIntakeClose());
-    driverGamepad.buttonPressed(BionicF310.Start, new InvertDriveDirection());
+    driverGamepad.buttonPressed(BionicF310.RT, 0.2, new InvertDriveDirection());
+
+    manipulatorGamepad.buttonPressed(BionicF310.A, new SetElevatorPosition(-27220,1));
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/Robot.java]: all of the numbers used in lines 86-93 should preferably in constants or at least commented well, because it's not very apparent at the moment

File: src/main/java/frc/team4909/robot/Robot.java [hunk 7]
```diff
@@ -40,43 +50,50 @@
...
   }
 
-  /**
+  /**''
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/Robot.java]: was this just a typo?

File: src/main/java/frc/team4909/robot/RobotConstants.java [hunk 8]
```diff
@@ -4,27 +4,25 @@
...
   public static final double speedMultiplier = .7;  
 
   /* Intake */
   public static final double cargoIntakeInSpeed = 1.0;
-  public static final double cargoIntakeHoldSpeed = 0;
+  public static final double cargoIntakeHoldSpeed = 0.0625;
   public static final double cargoIntakeOutSpeed = -1.0;
 
-  public static final double cargoIntakeCurrentLimit = 20;
+  public static final double cargoIntakeCurrentLimit = 0;
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/RobotConstants.java]: intake limit code needs to be fixed OR cherry picked out of this branch before it's merged.

File: src/main/java/frc/team4909/robot/RobotConstants.java [hunk 9]
```diff
@@ -4,27 +4,25 @@
...
 
   /* LIDAR */
   public static final int calibrationOffset = -2;
-  public static final double lidarLimit = 20; // cm
+  public static final double lidarLimit = 30; // cm
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/RobotConstants.java]: what does this number even mean?

File: src/main/java/frc/team4909/robot/subsystems/drivetrain/commands/Drive.java [hunk 10]
```diff
@@ -12,8 +12,8 @@ public Drive() {
     public void execute() {
         // Calls tank function using left Y and right Y
         // Negated since Y Axis is scaled -1 to +1 top to bottom (counterintuively)
-        Robot.drivetrainSubsystem.tankDrive(-Robot.driverGamepad.getThresholdAxis(BionicF310.LY),
-                -Robot.driverGamepad.getThresholdAxis(BionicF310.RY));
+        Robot.drivetrainSubsystem.arcadeDrive(-Robot.driverGamepad.getRawAxis(1),
+                -Robot.driverGamepad.getRawAxis(4));
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/drivetrain/commands/Drive.java]: why raw axis not threshold..? the reason we use threshold is so that we ignore the "deadband" around the center (resting is like 0.062) or something similar... if you truly would like raw axes, I'd recommend creating a method on BionicJoystick so we can pass in BionicF310.LY to that as well, or at the very least comment what axes here, because the comments on line 13 is outdated.

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java [hunk 11]
```diff
@@ -3,6 +3,7 @@
 import edu.wpi.first.wpilibj.Encoder;
 import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
 import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+// import com.sun.org.apache.xerces.internal.xni.QName;
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: uhhh this is definitely not supposed to be here

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java [hunk 12]
```diff
@@ -12,60 +13,63 @@
...
-        }
-    }
+        leftSPX.follow(leftSRX);  //All slaves will output the same value as the master SRX
+        rightSPX1.follow(leftSRX);  
+        rightSPX2.follow(leftSRX);
+        
+        rightSPX1.setInverted(true);
+        rightSPX2.setInverted(true);
+        leftSRX.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative);
+        leftSRX.selectProfileSlot(1, 0);
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: i would recommend just using slot 0 for all of the config, including the below PID constants

File: src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java [hunk 13]
```diff
@@ -1,19 +1,28 @@
 package frc.team4909.robot.subsystems.elevator.commands;
 
+import edu.wpi.first.wpilibj.command.Command;
 import edu.wpi.first.wpilibj.command.InstantCommand;
 import frc.team4909.robot.subsystems.elevator.ElevatorSubsystem;
 import frc.team4909.robot.Robot;
 
 public class SetElevatorPosition extends InstantCommand {
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java]: there's your issue. instant commands are designed to immediately exit, that would cause the OI control to take precedence

File: src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java [hunk 14]
```diff
@@ -1,19 +1,28 @@
...
-    public SetElevatorPosition(int setpoint) {
-        this.setpoint = setpoint;       
+    public SetElevatorPosition(int setpoint, int threshold) {
+        this.setpoint = setpoint;  
+        this.threshold = threshold;     
         requires(Robot.elevatorSubsystem);
     }
 
     @Override
     public void initialize() {
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java]: why in initialize? theoretically it shouldn't matter but I'd do it in execute to avoid being overwritten and to avoid any timeout issues that may be hidden

File: src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java [hunk 15]
```diff
@@ -1,19 +1,28 @@
...
-        Robot.elevatorSubsystem.holdingPosition = setpoint;
+        Robot.elevatorSubsystem.setPosition(setpoint);
+
+    }
+
+    @Override
+    protected boolean isFinished() {
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java]: this *may* fix the instant command thing but it's very hacky, might as well use the proper class from the start

File: src/main/java/frc/team4909/robot/subsystems/elevatorarm/ElevatorArmSubsystem.java [hunk 16]
```diff
@@ -11,31 +12,25 @@
...
+        elevatorArmSRX = new TalonSRX(RobotMap.elevatorArmSRXID);
 
         elevatorArmSRX.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative);
+ 
+        elevatorArmSRX.configContinuousCurrentLimit(3);
+        elevatorArmSRX.configPeakCurrentLimit(6);
+        
+        elevatorArmSRX.config_kP(0,1, 0);
+        elevatorArmSRX.config_kI(0,0);
+        elevatorArmSRX.config_kD(0,30, 0);
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevatorarm/ElevatorArmSubsystem.java]: Robot Constants here too

**Comment by roshanr10** [comment]
I've made some comments, but for future reference, this is a significant amount of code within a given PR, and a majority of the changes come from differing issues. This makes it difficult to test isolated aspects of the changes

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java (hunk 2 above)
Comment (ashwinc12) [reply src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: Fixed

**Comment by ashwinc12** [comment]
@roshanr10 @derrell Can one of you fix these conflicts and merge into testing? It doesn't seem to be allowing me to open in desktop for some reason and I'm not that familiar with command line

**Comment by roshanr10** [comment]
I won’t have access my laptop until late at night— if it’s not merged by
then, I can fix it

**Comment by ashwinc12** [comment]
@roshanr10 Sweet thank you.

**Comment by roshanr10** [comment]
@ashwinc12 done, let me know if you need anything else... was intake limit dropped for now?

**Comment by roshanr10** [comment]
@ashwinc12 do you want me to review this code, or are there other pending changes you want me to wait for?

**Comment by roshanr10** [comment]
@ashwinc12 @jasonbishop11 @ShashJar can we get an itemized list of what issues this PR addresses? this is a decent chunk of code, and we should make sure we hit each aspect we meant to cover

**Comment by ashwinc12** [comment]
@ShashJar Can you get that done? ^

**Comment by ashwinc12** [comment]
Intake current limit hasn't been played with since you left so it has been neglected for the most part. As for the code, I reviewed it earlier today and fixed the changes that I saw but an extra pair of eyes never hurts. Ideally merged to testing and released before Monday's meeting.

**Comment by roshanr10** [comment]
okay I'll take a quick peek, also have IR sensors been mounted and tested?

File: src/main/java/frc/team4909/robot/RobotConstants.java (hunk 9 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/RobotConstants.java]: @ShashJar comments please

File: src/main/java/frc/team4909/robot/Robot.java [hunk 17]
```diff
@@ -107,14 +170,36 @@ public void autonomousPeriodic() {
...
+    if (CargoIntake == true){
+      new CargoIntakeIn();
+      SmartDashboard.putBoolean("Cargo In", false);
+    }
+
+    CargoOuttake = SmartDashboard.getBoolean("Cargo Out", false);
+    if (CargoOuttake == true){
+      new CargoIntakeOut();
+      SmartDashboard.putBoolean("Cargo Out", false);
+    }
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/Robot.java]: what are these SmartDashboard references for? @ashwinc12

Comment (roshanr10) [reply src/main/java/frc/team4909/robot/Robot.java]: is this intended to be a way to run these commands from smartdashboard? this should only construct the commands, it doesn't appear to be passing it to the scheduler

File: src/main/java/frc/team4909/robot/subsystems/climber/ClimberSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+import frc.team4909.robot.subsystems.drivetrain.commands.Drive;
+import frc.team4909.robot.subsystems.climber.commands.ExtendStilts;
+import frc.team4909.robot.subsystems.climber.commands.RetractStilts;
+import com.ctre.phoenix.motorcontrol.FeedbackDevice;
+import edu.wpi.first.wpilibj.command.CommandGroup;
+import frc.team4909.robot.Robot;
+import frc.team4909.robot.operator.controllers.BionicF310;
+
+public class ClimberSubsystem extends Subsystem {
+    WPI_TalonSRX climberSRX;
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/climber/ClimberSubsystem.java]: this naming needs some work, that's indecipherable anywhere throughout the code. take the lines 25 & 26 and apply them here

ex. `climberLiftSRX` and `climberDriveSPX`

File: src/main/java/frc/team4909/robot/subsystems/climber/ClimberSubsystem.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    WPI_VictorSPX climberSPX;
+
+    public ClimberSubsystem() {
+        climberSRX = new WPI_TalonSRX(RobotMap.climberSRXID); // Climber lift
+        climberSPX = new WPI_VictorSPX(RobotMap.climberSPXID); // CLimber drive
+
+        climberSRX.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative);
+
+        /* PID yet to be calibrated */
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/climber/ClimberSubsystem.java]: personal thing, but I like to prefix all of my remaining work comments with `TODO: ` so that it's easy to check if you finished everything by doing the file search across the workspace... either way, I'd make sure this gets tuned tomorrow(errr i guess today now) if possible

File: src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java (hunk 11 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/subsystems/elevator/ElevatorSubsystem.java]: ^^

File: src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java (hunk 15 above)
Comment (roshanr10) [reply src/main/java/frc/team4909/robot/subsystems/elevator/commands/SetElevatorPosition.java]: ^^ there should be an isfinished condition so it can be strung together in a semiauto

File: src/main/java/frc/team4909/robot/subsystems/elevatorarm/commands/SetAngle.java [hunk 20]
```diff
@@ -16,4 +16,5 @@ public SetAngle(int setpoint, ElevatorArmSubsytem elevatorArm) {
     public void initialize() {
         elevatorArm.holdingPosition = setpoint;
     }
+
```
Comment (roshanr10) [inline src/main/java/frc/team4909/robot/subsystems/elevatorarm/commands/SetAngle.java]: same as above with the exit condition– this idea worked well last year but ideally the command should only exit when it's within an acceptable threshold and the action can be considered "complete"

File: src/main/java/frc/team4909/robot/RobotConstants.java (hunk 9 above)
Comment (ShashJar) [reply src/main/java/frc/team4909/robot/RobotConstants.java]: Lidar sensor isn't placed at the exact edge of the robot, therefore we need to account for the width of the bumper/space between bumper and lidar - we measured this to be ~9 cm.

**Comment by roshanr10** [comment]
Seeing some late changes, has ALL of the changes been tested? Just curious,
trying to get a bearing on things. Anything I can help with?

**Comment by ashwinc12** [comment]
Yes all of the code on the branch was tested code today. This needed to be merged prior to tomorrow's meeting because it was a big problem having everyone working on the same branch.

**Comment by roshanr10** [comment]
I take it you saw Shash’s last minute commits? Just want to confirm you
intended to merge that too.

**Comment by ashwinc12** [comment]
Yeah I ended up deleting everything from his last commit and asked him to branch off once I merged into testing.

**Comment by roshanr10** [comment]
Got it, didn't see that last commit – all good then. 

For future reference, this feature does exist in git: https://www.atlassian.com/git/tutorials/undoing-changes/git-revert
